File: web/packages/helix-backend/src/device-auth/offline-code.ts
import { createHmac, timingSafeEqual } from 'node:crypto';

/**
 * The cloud half of the offline challenge/response.
 *
 * A device that cannot reach the cloud shows a challenge; the user, on a phone
 * that can, signs in and submits it; this produces the response they type back.
 * The response proves only that the cloud authorized *this* user for *this*
 * device and *this* challenge — it carries no scopes, so it cannot grant
 * anything the device did not already know about the user.
 *
 * The device recomputes the same digest locally. Both implementations are pinned
 * by `linux/device/go/internal/authd/testdata/offline-response-vectors.json`, and
 * a change here that is not mirrored in Go will fail that suite.
 */

/** Omits I, L, O, 0 and 1: these codes are read aloud and retyped by hand. */
const CODE_ALPHABET = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
const CODE_RAW_BYTES = 5;
const CODE_CHARS = 8;
const CODE_GROUP = 4;
/** A uid is written as an unsigned 64-bit big-endian integer. */
const UID_BYTES = 8;

/** Separates these digests from any other use of the same device secret. */
const OFFLINE_DOMAIN = 'HXOFF1';

const BITS_PER_CHAR = 5n;
const BITS_PER_BYTE = 8n;
const CHAR_MASK = 0x1fn;
const BYTE_MASK = 0xffn;

/** Everything a response is bound to. Change any field and the response changes. */
export type OfflineBinding = Readonly<{
  deviceId: string;
  userId: string;
  linuxUid: number;
  challenge: string;
}>;

/** Canonical form: upper case, no separators. Internal: the exported helpers all
 * normalize their own input, so callers never need to. */
const normalizeCode = (code: string): string =>
  code.trim().toUpperCase().replaceAll('-', '').replaceAll(' ', '');

/** Display form, hyphenated for reading aloud. */
export const formatCode = (code: string): string => {
  const normalized = normalizeCode(code);
  if (normalized.length !== CODE_CHARS) {
    return normalized;
  }
  return `${normalized.slice(0, CODE_GROUP)}-${normalized.slice(CODE_GROUP)}`;
};

/** Renders five bytes as an eight-character code. */
export const encodeCode = (raw: Uint8Array): string => {
  if (raw.length !== CODE_RAW_BYTES) {
    throw new RangeError(`offline code needs ${CODE_RAW_BYTES} bytes, got ${raw.length}`);
  }
  let packed = 0n;
  for (const byte of raw) {
    packed = (packed << BITS_PER_BYTE) | BigInt(byte);
  }

  let out = '';
  for (let i = 0; i < CODE_CHARS; i++) {
    const shift = BITS_PER_CHAR * BigInt(CODE_CHARS - 1 - i);
    out += CODE_ALPHABET[Number((packed >> shift) & CHAR_MASK)];
  }
  return out;
};

/** Parses a code back to its bytes, accepting any case and the display hyphen. */
export const decodeCode = (code: string): Uint8Array => {
  const normalized = normalizeCode(code);
  if (normalized.length !== CODE_CHARS) {
    throw new RangeError(`offline code must be ${CODE_CHARS} characters, got ${normalized.length}`);
  }

  let packed = 0n;
  for (const char of normalized) {
    const index = CODE_ALPHABET.indexOf(char);
    if (index < 0) {
      throw new RangeError(`offline code contains ${char}, which is not in the alphabet`);
    }
    packed = (packed << BITS_PER_CHAR) | BigInt(index);
  }

  const out = new Uint8Array(CODE_RAW_BYTES);
  for (let i = CODE_RAW_BYTES - 1; i >= 0; i--) {
    out[i] = Number(packed & BYTE_MASK);
    packed >>= BITS_PER_BYTE;
  }
  return out;
};

/** A 16-bit big-endian length followed by the bytes, so adjacent fields in the
 * digest can never run together. */
const lengthPrefixed = (value: string): Buffer => {
  const bytes = Buffer.from(value, 'utf8');
  const prefix = Buffer.alloc(2);
  prefix.writeUInt16BE(bytes.length);
  return Buffer.concat([prefix, bytes]);
};

/** Computes the response a user should type back into the device. */
export const offlineResponse = (secret: Buffer, binding: OfflineBinding): string => {
  if (secret.length === 0) {
    throw new RangeError('offline device secret is empty');
  }
  const challenge = decodeCode(binding.challenge);

  const uid = Buffer.alloc(UID_BYTES);
  uid.writeBigUInt64BE(BigInt(binding.linuxUid));

  const digest = createHmac('sha256', secret)
    .update(Buffer.from(OFFLINE_DOMAIN, 'utf8'))
    .update(lengthPrefixed(binding.deviceId))
    // The device's user id is text, so it is length-prefixed like the device id.
    .update(lengthPrefixed(binding.userId))
    .update(uid)
    .update(challenge)
    .digest();

  return encodeCode(new Uint8Array(digest.subarray(0, CODE_RAW_BYTES)));
};

/** Compares two codes without leaking where they differ. */
export const codesMatch = (expected: string, given: string): boolean => {
  const a = Buffer.from(normalizeCode(expected), 'utf8');
  const b = Buffer.from(normalizeCode(given), 'utf8');
  if (a.length !== b.length || a.length === 0) {
    return false;
  }
  return timingSafeEqual(a, b);
};

File: linux/device/go/internal/authd/offlinecode.go
// SPDX-License-Identifier: AGPL-3.0-only

package authd

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"strings"
)

// The offline challenge/response wire format.
//
// A person reads a code off a screen, types it into a phone, reads another code
// back, and types that into a terminal. Everything here follows from that: the
// alphabet has no characters that look alike, the codes are short enough to
// retype without resentment, and the whole exchange is bound to one device, one
// user and one challenge so a code is worthless anywhere else.
const (
	// codeAlphabet omits I, L, O, 0 and 1: they are indistinguishable in most
	// fonts and this is read aloud and retyped by hand.
	codeAlphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
	// codeRawBytes is 40 bits, which encodes to exactly codeChars characters.
	codeRawBytes = 5
	codeChars    = 8
	// codeGroup is where the display hyphen goes: K7MP-Q29X.
	codeGroup = 4

	// offlineDomain separates these digests from any other use of the same key.
	offlineDomain = "HXOFF1"
)

// encodeCode renders raw bytes as an 8-character code.
func encodeCode(raw []byte) (string, error) {
	if len(raw) != codeRawBytes {
		return "", fmt.Errorf("offline code needs %d bytes, got %d", codeRawBytes, len(raw))
	}

	var packed uint64
	for _, b := range raw {
		packed = packed<<8 | uint64(b)
	}

	out := make([]byte, codeChars)
	for i := codeChars - 1; i >= 0; i-- {
		out[i] = codeAlphabet[packed&0x1F]
		packed >>= 5
	}
	return string(out), nil
}

// decodeCode parses a code back to its bytes, accepting what a human is likely
// to type: any case, with or without the display hyphen or spaces.
func decodeCode(code string) ([]byte, error) {
	normalized := normalizeCode(code)
	if len(normalized) != codeChars {
		return nil, fmt.Errorf("offline code must be %d characters, got %d", codeChars, len(normalized))
	}

	var packed uint64
	for _, char := range normalized {
		index := strings.IndexRune(codeAlphabet, char)
		if index < 0 {
			return nil, fmt.Errorf("offline code contains %q, which is not in the alphabet", char)
		}
		packed = packed<<5 | uint64(index)
	}

	out := make([]byte, codeRawBytes)
	for i := codeRawBytes - 1; i >= 0; i-- {
		out[i] = byte(packed & 0xFF)
		packed >>= 8
	}
	return out, nil
}

// normalizeCode is the canonical form: upper case, no separators.
func normalizeCode(code string) string {
	var b strings.Builder
	for _, char := range strings.ToUpper(strings.TrimSpace(code)) {
		if char == '-' || char == ' ' {
			continue
		}
		b.WriteRune(char)
	}
	return b.String()
}

// formatCode is the display form, hyphenated for reading aloud.
func formatCode(code string) string {
	normalized := normalizeCode(code)
	if len(normalized) != codeChars {
		return normalized
	}
	return normalized[:codeGroup] + "-" + normalized[codeGroup:]
}

// OfflineBinding is everything a response is tied to. Change any field and the
// response changes, which is what stops one person's code working for another.
type OfflineBinding struct {
	DeviceID  string
	UserID    string
	LinuxUID  uint32
	Challenge string
}

// OfflineResponse computes the response the cloud would issue for a binding.
//
// The device recomputes this locally and compares, which is the whole trick: no
// network is involved, yet the code can only have come from something holding
// both the device secret and an authorization for that exact user.
func OfflineResponse(secret []byte, binding OfflineBinding) (string, error) {
	challenge, err := decodeCode(binding.Challenge)
	if err != nil {
		return "", err
	}
	if len(secret) == 0 {
		return "", fmt.Errorf("offline device secret is empty")
	}

	mac := hmac.New(sha256.New, secret)
	mac.Write([]byte(offlineDomain))
	writeLengthPrefixed(mac, binding.DeviceID)
	// The spec models the user as a 16-byte UUID. A Helix user id is text, so it
	// is length-prefixed like the device id: same binding, correct type.
	writeLengthPrefixed(mac, binding.UserID)
	_ = binary.Write(mac, binary.BigEndian, uint64(binding.LinuxUID))
	mac.Write(challenge)

	return encodeCode(mac.Sum(nil)[:codeRawBytes])
}

// writeLengthPrefixed writes a 16-bit big-endian length then the bytes, so two
// adjacent fields can never be confused for one another.
func writeLengthPrefixed(w interface{ Write([]byte) (int, error) }, value string) {
	_ = binary.Write(w, binary.BigEndian, uint16(len(value)))
	_, _ = w.Write([]byte(value))
}
